### PR TITLE
Feature: Added fts_throttler #2482

### PR DESCRIPTION
**Pull request by DilaksunB**
Feature: Added fts_throttler #2482

### CODE REVIEW

File: lib/rucio/daemons/conveyor/fts_throttler.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+                cycle_file = config_get('conveyor', 'fts_throttler_cycle')
+            except Exception:
+                logging.warn('could not get the cycle file, cannot perform tuning for this cycle without cycle file, returning')
+                return
+
+            rses = result['aggregations']['rse']['buckets']
+            cycle_info_dict = {'storages': []}
+            for rse in rses:
+                # if a rse has a failure ratio above 25% we tune it.
+                if rse['failure_ratio'].get('value') > 25:
```
Comment (cserf) [inline lib/rucio/daemons/conveyor/fts_throttler.py]: The value should be tunable. You can make it as an option of rucio config

File: lib/rucio/daemons/conveyor/fts_throttler.py [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+                    # rse_info holds the storage name(0) and FTS-host server(1)
+                    rse_info = rse['key'].split()
+
+                    # Tapes might have other reasons for timeouts which should be treated differently, therefor they are ignored and not tuned for now.
+                    if rse_info[0].endswith('TAPE'):
```
Comment (cserf) [inline lib/rucio/daemons/conveyor/fts_throttler.py]: The check relies on the naming convention of the RSE. In ATLAS we have some TAPE RSEs that do not ends with `TAPE`. A best check is to check the `rse_type` which is either `DISK` or `TAPE`

File: lib/rucio/daemons/conveyor/fts_throttler.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+                    # instantiate transfertool for access to get_se_config and set_se_config.
+                    t = FTS3Transfertool(rse_info[1])
+
+                    # extract FTS storage from dst-url
+                    tmp = rse['destination']['hits']['hits'][0]['_source']['payload']['dst-url'].split(':', 2)
+                    url = tmp[0] + ':' + tmp[1]
+
+                    n = rse['failure_ratio'].get('value')
+
+                    print('\033[91m RSE \033[0m' + rse_info[0] + '\033[91m on FTS host \033[0m' + rse_info[1] +
```
Comment (cserf) [inline lib/rucio/daemons/conveyor/fts_throttler.py]: Remove the print statement

File: lib/rucio/daemons/conveyor/fts_throttler.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+                                     ', outbound_max_active changed from ' + str(default_storage['se_info']['outbound_max_active']) + ' to ' + str(int((100 / (100 + n)) * default_storage['se_info']['outbound_max_active'])))
+
+                        # cycle_info_dict is used to write changes down to the cycle file.
+                        cycle_info_dict['storages'].append({'storage': url, 'inbound_max_active': default_storage['se_info']['inbound_max_active'],
+                                                            'outbound_max_active': default_storage['se_info']['outbound_max_active'], 'failure_ratio': n,
+                                                            'tuned_inbound_max_active': int((100 / (100 + n)) * default_storage['se_info']['inbound_max_active']),
+                                                            'tuned_outbound_max_active': int((100 / (100 + n)) * default_storage['se_info']['outbound_max_active']),
+                                                            'fts-host': rse_info[1], 'time': str(datetime.datetime.now())})
+                        continue
+                    except Exception, error:
```
Comment (cserf) [inline lib/rucio/daemons/conveyor/fts_throttler.py]: Not py3k compatible : https://python-future.org/compatible_idioms.html#catching-exceptions

File: lib/rucio/daemons/conveyor/fts_throttler.py [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+        Read the failure ratio of storages without tuning
+        :returns: filtered JSON response from Elastic search.
+        """
+        result = self.request_timeout_data()
+        if result is not None:
+
+            rses = result['aggregations']['rse']['buckets']
+            for rse in rses:
+                # if a rse has a failure ratio above 25% we tune it.
+                if rse['failure_ratio'].get('value') > 25:
```
Comment (cserf) [inline lib/rucio/daemons/conveyor/fts_throttler.py]: Same comment as above

File: lib/rucio/daemons/conveyor/fts_throttler.py [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+                    tmp = rse['destination']['hits']['hits'][0]['_source']['payload']['dst-url'].split(':', 2)
+                    url = tmp[0] + ':' + tmp[1]
+                    print('\033[91m RSE \033[0m' + rse_info[0] + '\033[91m on FTS host \033[0m' + rse_info[1] + '\033[91m has failure ratio \033[0m' + str(rse['failure_ratio'].get('value')) + '\033[91m on storage \033[0m' + url)
+
+                    try:
+                        se = t.get_se_config(url)
+                        print(se)
+                    except KeyError:
+                        logging.warn('configuration for storage element was not found')
+                    except Exception, error:
```
Comment (cserf) [inline lib/rucio/daemons/conveyor/fts_throttler.py]: Non py3k compatible. See above

**Review by cserf — COMMENTED** [review]
First comments after a quick look. Will have a more detailed one.

File: bin/rucio-conveyor-fts_throttler [hunk 7]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python
```
Comment (bari12) [inline bin/rucio-conveyor-fts_throttler]: Please rename to `rucio-conveyor-fts-throttler`
(no underscore)

**Review by bari12 — DISMISSED** [review]

**Review by cserf — APPROVED** [review]

**Pull request merged**